sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -441,6 +456,25 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
 
       case Not(IsNull(e)) => IsNotNull(e)
       case Not(IsNotNull(e)) => IsNull(e)
+
+      // Move `Not` from one side of `EqualTo`/`EqualNullSafe` to the other side if it's beneficial.
+      // E.g. `EqualTo(Not(a), b)` where `b = Not(c)`, it will become
+      // `EqualTo(a, Not(b))` => `EqualTo(a, Not(Not(c)))` => `EqualTo(a, c)`
+      // In addition, `if canSimplifyNot(b)` checks if the optimization can converge
+      // that avoids the situation two conditions are returning to each other.
+      case EqualTo(Not(a), b) if canSimplifyNot(b) => EqualTo(a, Not(b))
```

Comment:
Do we want to add the checks like the following? This will help that we apply all the `Not` optimization first before we hit the following roles, and also, the checks explicitly ensure the roles will converge.
```
      case EqualTo(Not(a), b) if !canSimplifyNot(a) && canSimplifyNot(b) => EqualTo(a, Not(b))
      case EqualTo(a, Not(b)) if canSimplifyNot(a) && !canSimplifyNot(b) => EqualTo(Not(a), b)
```

Reply:
We can but the extra checks are a bit redundant because all `Not` optimizations are listed above this line of the case match. Do you feel we must need the extra checks? @dbtsai 
It is actually does not matter `a` or `b` gets simplified because `canSimplifyNot()` promises it will be simpler than current form.

Reply:
Added the checks